### ISSUE TITLE
Cordon node with a do-not-evict pod when ttlSecondsUntilExpired is met

### ISSUE THREAD

**Issue by cest-pas-faux** [kubernetes-sigs/karpenter#622]
### Tell us about your request

Regarding the do-not-evict annotation, it currrently prevents some nodes to be deprovisioned on our clusters.

I know, that's the goal and i'm fine with it, but would it be possible to cordon the node at least when the `ttlSecondsUntilExpired` is met ?

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

We provision expensive nodes as a failback for some workloads with a ttl of 3 hours, and we saw some still running with a 6 hours uptime.
The annotation was preventing the drain/deletion as cronjobs kept being scheduled on it, but a manual cordon fixed it within minutes.

### Are you currently working around this issue?

No

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jukie** [comment]
Has there been any more discussions on this? Would PRs adding this behavior be accepted?

**Comment by jonathan-innis** [comment]
> Has there been any more discussions on this

I think we should consider discussing this in the next working group. I'd like to see a design on this one considering the impact that this could have to the cluster, given there could be cases if we're not careful where all nodes on the cluster get tainted all at once. 

As an example, if we have a set of nodes that are expired all at once and we cordon them all at once, is that a reasonable stance to take, knowing that those nodes will get eventually disrupted? Probably, assuming that our disruption logic can handle the disruption quick enough that have underutilization on that node isn't a problem, but what if that node has a fully-blocking PDB? What if that node has a `do-not-evict` pod and that pod never goes away? Do we need some forceful termination that happens after a node has been cordoned with a blocking pod for some amount of time? There might be some interactions here that we should consider with #743

**Comment by jukie** [comment]
Agreed there are some risks and also some relation to other proposed features under https://github.com/aws/karpenter/issues/1738 but I don't think that should limit from implementing this through a feature gate or by disabling as default.

How do I see the schedule for bi-weekly's? Looks like the notes haven't been updated in awhile so it's unclear whether it's this coming week or next.

**Comment by njtran** [comment]
@jukie sorry for the unfortunate timing, as I just came off vacation. Just noticed you asked here, but we just had a WG an hour ago. You can check when and where these meetings are happening here: https://github.com/kubernetes-sigs/karpenter#community-discussion-contribution-and-support. You can message me or @jonathan-innis in the kubernetes slack as well if you want to discuss more.

**Comment by jukie** [comment]
Thanks, I'll reach out and can discuss further there!

**Comment by mseiwald** [comment]
Is there any update on this? We are interested in this feature specifically for K8s upgrade situations where we do not want new pods to be scheduled to outdated nodes even if the node contains pods with the do-not-disrupt annotation.

**Comment by sftim** [comment]
> Is there any update on this?

So far as I know, the issue history tracks progress here and isn't missing any important details.

**Comment by ugurgural** [comment]
Hi, any updates on here? Last time I heard [on this from](https://github.com/aws/karpenter-provider-aws/issues/5664) @jonathan-innis that this is still on debate.

**Comment by jukie** [comment]
Does the `karpenter.sh/disrupted:NoSchedule` provide this already @ugurgural?

Ref: https://karpenter.sh/docs/concepts/disruption/#termination-controller

That taint should be applied once a node becomes expired or marked for disruption, is there a use case for specifically cordoning the node as well?

**Comment by sftim** [comment]
> `karpenter.sh/disrupted:NoSchedule`

A more specific taint feels appropriate. eg `karpenter.sh/node-ttl-expired:NoSchedule` and eventually `karpenter.sh/node-ttl-expired:NoExecute`

**Comment by mseiwald** [comment]
> Does the `karpenter.sh/disrupted:NoSchedule` provide this already @ugurgural?
> 
> Ref: https://karpenter.sh/docs/concepts/disruption/#termination-controller
> 
> That taint should be applied once a node becomes expired or marked for disruption, is there a use case for specifically cordoning the node as well?

My understanding is that it will never get to the stage where it would add the taint `karpenter.sh/disrupted:NoSchedule`.

> If there are [pods that cannot be evicted](https://karpenter.sh/docs/concepts/disruption/#pod-eviction) on the node, Karpenter will ignore the node and try disrupting it later.

https://karpenter.sh/docs/concepts/disruption/#disruption-controller

If a node contains Pods with the `do-not-disrupt` annotation it will never trigger the disruption and therefore never add this taint.

I think this feature request is about changing this flow to add the taint (cordoning is in the end similar to tainting a node) even if there are Pods with `do-not-disrupt` annotation so that they will not be scheduled to the outdated nodes on next Pod restart.

**Comment by jonathan-innis** [comment]
This is fixed now with the [forceful disruption feature](https://github.com/kubernetes-sigs/karpenter/blob/main/designs/forceful-expiration.md) that was pushed out in v1.0. Note that `karpenter.sh/do-not-disrupt` pods will still block the draining of the node, but it won't block the node from beginning getting drained. If you then also wanted to force the _removal_ of the node after some amount of time, you can use `terminationGracePeriod` in tandem with `expireAfter`.

**Comment by mseiwald** [comment]
> This is fixed now with the [forceful disruption feature](https://github.com/kubernetes-sigs/karpenter/blob/main/designs/forceful-expiration.md) that was pushed out in v1.0. Note that `karpenter.sh/do-not-disrupt` pods will still block the draining of the node, but it won't block the node from beginning getting drained. If you then also wanted to force the _removal_ of the node after some amount of time, you can use `terminationGracePeriod` in tandem with `expireAfter`.

Hi @jonathan-innis! This feature only affects expiration but not drift, correct? In my experience drifted nodes will not even start the disruption process when they still have do-not-disrupt Pods on them which is also what [other people](https://github.com/kubernetes-sigs/karpenter/issues/623) are reporting.
